community/kernel/src/main/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanStore.java

```diff
@@ -0,0 +1,265 @@
+     */
+    private final int pageSize;
+
+    /**
+     * The index which backs this label scan store. Instantiated in {@link #init()} and considered
+     * started after call to {@link #start()}.
+     */
+    private GBPTree<LabelScanKey,LabelScanValue> index;
+
+    /**
+     * Whether or not {@link #start()} has been called.
+     */
+    private boolean started;
+
+    /**
+     * If {@link #index} is {@code null} and {@link #started} is {@code false},
+     * then it's between {@link #init()} and {@link #start()}. If {@link #newWriter()} is called at this
+     * point we infer that recovery is taking place and so we notify the {@link GBPTree} about that fact.
+     */
+    private boolean recoveryStarted;
+
+    public NativeLabelScanStore( PageCache pageCache, File storeDir, int rangeSize, int pageSize,
+            FullStoreChangeStream fullStoreChangeStream )
+    {
+        this.pageCache = pageCache;
+        this.pageSize = pageSize;
+        this.fullStoreChangeStream = fullStoreChangeStream;
+        this.storeFile = new File( storeDir, DEFAULT_NAME + ".labelscanstore.db" );
+        this.rangeSize = rangeSize;
+    }
+
+    /**
+     * @return {@link LabelScanReader} capable of finding node ids with given label ids.
+     * Readers will immediately see updates made by {@link LabelScanWriter}, although {@link LabelScanWriter}
+     * may internally batch updates so functionality isn't realiable. The only given is that readers will
+     * see at least updates from closed {@link LabelScanWriter writers}.
+     */
+    @Override
+    public LabelScanReader newReader()
+    {
+        return new NativeLabelScanReader( index, rangeSize );
+    }
+
+    /**
+     * Returns {@link LabelScanWriter} capable of making changes to this {@link LabelScanStore}.
+     * Only a single writer is allowed at any given point in time.
+     *
+     * @param batching whether or not the returned writer will be used for batch-insertion where
+     * updates, usually a large amount, will arrive in sorted order.
+     * @return {@link LabelScanWriter} capable of making changes to this {@link LabelScanStore}.
+     * @throws IllegalStateException if someone else has already acquired a writer and hasn't yet
+     * called {@link LabelScanWriter#close()}.
+     */
+    @Override
+    public LabelScanWriter newWriter( boolean batching )
+    {
+        try
+        {
+            if ( !started && !recoveryStarted )
+            {
+                // Let's notify our index that recovery is about to commence, we do this once before
+                // the first recovered transaction gets applied.
+                index.prepareForRecovery();
+                recoveryStarted = true;
+            }
+
+            IndexWriter<LabelScanKey,LabelScanValue> inserter = index.writer();
+            return new NativeLabelScanWriter( inserter, rangeSize, batching ? 10_000 : 1_000 );
```

Comment:
Initially this `batching` meant more, but now it changes only the internal batch size. We'll measure if this really matters at all, otherwise revert this argument addition.